File: api-python/budget.py
"""Hard daily spend cap and per-IP limit. Mirrors api/src/budget.ts exactly."""
from __future__ import annotations

import hashlib
import os
import time
from datetime import datetime, timezone

import boto3
from botocore.exceptions import ClientError

TABLE = os.environ.get("USAGE_TABLE", "")
CAP = float(os.environ.get("DAILY_BUDGET_USD", "5"))
PRICE_IN = float(os.environ.get("PRICE_IN_PER_MTOK", "3"))
PRICE_OUT = float(os.environ.get("PRICE_OUT_PER_MTOK", "15"))
RESERVE = float(os.environ.get("RESERVE_USD", "0.04"))
IP_TURNS_PER_HOUR = int(os.environ.get("IP_TURNS_PER_HOUR", "40"))
OVERHEAD = 0.0005

_ddb = boto3.client("dynamodb")


def day_key() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")


def turn_cost(usage: dict) -> float:
    return usage.get("inputTokens", 0) / 1e6 * PRICE_IN + usage.get("outputTokens", 0) / 1e6 * PRICE_OUT + OVERHEAD


def _conditional_failed(err: ClientError) -> bool:
    return err.response.get("Error", {}).get("Code") == "ConditionalCheckFailedException"


def reserve(ip: str | None) -> dict:
    """Returns {"ok": True, "day": ...} or {"ok": False, "status": 429, "message": ...}."""
    if not TABLE:
        return {"ok": True, "day": day_key()}
    day = day_key()
    if ip:
        hour = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H")
        ip_key = f"ip#{hashlib.sha256(ip.encode()).hexdigest()[:16]}#{hour}"
        try:
            _ddb.update_item(TableName=TABLE, Key={"pk": {"S": ip_key}}, UpdateExpression="ADD turns :one SET #ttl = :ttl",
                             ConditionExpression="attribute_not_exists(turns) OR turns < :max", ExpressionAttributeNames={"#ttl": "ttl"},
                             ExpressionAttributeValues={":one": {"N": "1"}, ":max": {"N": str(IP_TURNS_PER_HOUR)}, ":ttl": {"N": str(int(time.time()) + 7200)}})
        except ClientError as err:
            if _conditional_failed(err):
                return {"ok": False, "status": 429, "message": f"That's a lot of questions from one place. This demo allows {IP_TURNS_PER_HOUR} an hour per visitor; try again a little later."}
            raise
    try:
        _ddb.update_item(TableName=TABLE, Key={"pk": {"S": f"day#{day}"}}, UpdateExpression="ADD spent :r, turns :one",
                         ConditionExpression="attribute_not_exists(spent) OR spent <= :limit",
                         ExpressionAttributeValues={":r": {"N": str(RESERVE)}, ":one": {"N": "1"}, ":limit": {"N": str(CAP - RESERVE)}})
        return {"ok": True, "day": day}
    except ClientError as err:
        if _conditional_failed(err):
            return {"ok": False, "status": 429, "message": f"This demo has used its ${CAP:g} budget for today. It resets at midnight UTC."}
        raise


def settle(day: str, actual_usd: float) -> float:
    if not TABLE:
        return 0.0
    res = _ddb.update_item(TableName=TABLE, Key={"pk": {"S": f"day#{day}"}}, UpdateExpression="ADD spent :d",
                           ExpressionAttributeValues={":d": {"N": f"{actual_usd - RESERVE:.6f}"}}, ReturnValues="ALL_NEW")
    return float(res.get("Attributes", {}).get("spent", {}).get("N", 0))


def status() -> dict:
    day = day_key()
    spent = turns = 0.0
    if TABLE:
        item = _ddb.get_item(TableName=TABLE, Key={"pk": {"S": f"day#{day}"}}).get("Item", {})
        spent = float(item.get("spent", {}).get("N", 0)); turns = float(item.get("turns", {}).get("N", 0))
    return {"day": day, "spent": max(0.0, spent), "cap": CAP, "remaining": max(0.0, CAP - spent), "turns": int(turns)}

File: api/src/budget.ts
// Hard daily spend cap and per-IP fairness limit, enforced before every turn.
// Bedrock has no spend limit of its own, so the cap lives here: one DynamoDB
// item per UTC day holds the running estimated spend; a turn first reserves an
// estimated cost with a conditional update (atomic, so concurrent turns cannot
// overshoot by more than their own reservations), then settles to the actual
// token cost when the turn finishes.

import { DynamoDBClient, UpdateItemCommand, GetItemCommand, ConditionalCheckFailedException } from "@aws-sdk/client-dynamodb";
import { createHash } from "node:crypto";

const TABLE = process.env.USAGE_TABLE ?? "";
const CAP = Number(process.env.DAILY_BUDGET_USD ?? 5);
const PRICE_IN = Number(process.env.PRICE_IN_PER_MTOK ?? 3);
const PRICE_OUT = Number(process.env.PRICE_OUT_PER_MTOK ?? 15);
const RESERVE = Number(process.env.RESERVE_USD ?? 0.04);
const IP_TURNS_PER_HOUR = Number(process.env.IP_TURNS_PER_HOUR ?? 40);
const OVERHEAD = 0.0005; // Lambda + CloudFront + DynamoDB per turn, generous

const ddb = new DynamoDBClient({});

export const dayKey = (d = new Date()) => d.toISOString().slice(0, 10);

export function turnCost(usage: { inputTokens: number; outputTokens: number }): number {
  return (usage.inputTokens / 1e6) * PRICE_IN + (usage.outputTokens / 1e6) * PRICE_OUT + OVERHEAD;
}

export type Reservation = { ok: true; day: string } | { ok: false; status: number; message: string };

export async function reserve(ip: string | undefined): Promise<Reservation> {
  if (!TABLE) return { ok: true, day: dayKey() };
  const day = dayKey();
  // Per-IP limit first: one bot must not spend everyone's day in three minutes.
  if (ip) {
    const hour = new Date().toISOString().slice(0, 13);
    const ipKey = `ip#${createHash("sha256").update(ip).digest("hex").slice(0, 16)}#${hour}`;
    try {
      await ddb.send(new UpdateItemCommand({
        TableName: TABLE, Key: { pk: { S: ipKey } },
        UpdateExpression: "ADD turns :one SET #ttl = :ttl",
        ConditionExpression: "attribute_not_exists(turns) OR turns < :max",
        ExpressionAttributeNames: { "#ttl": "ttl" },
        ExpressionAttributeValues: { ":one": { N: "1" }, ":max": { N: String(IP_TURNS_PER_HOUR) }, ":ttl": { N: String(Math.floor(Date.now() / 1000) + 7200) } },
      }));
    } catch (err) {
      if (err instanceof ConditionalCheckFailedException) return { ok: false, status: 429, message: `That's a lot of questions from one place. This demo allows ${IP_TURNS_PER_HOUR} an hour per visitor; try again a little later.` };
      throw err;
    }
  }
  try {
    await ddb.send(new UpdateItemCommand({
      TableName: TABLE, Key: { pk: { S: `day#${day}` } },
      UpdateExpression: "ADD spent :r, turns :one",
      ConditionExpression: "attribute_not_exists(spent) OR spent <= :limit",
      ExpressionAttributeValues: { ":r": { N: String(RESERVE) }, ":one": { N: "1" }, ":limit": { N: String(CAP - RESERVE) } },
    }));
    return { ok: true, day };
  } catch (err) {
    if (err instanceof ConditionalCheckFailedException) return { ok: false, status: 429, message: `This demo has used its $${CAP} budget for today. It resets at midnight UTC.` };
    throw err;
  }
}

/** Replace the reservation with the real cost. Returns today's running total. */
export async function settle(day: string, actualUsd: number): Promise<number> {
  if (!TABLE) return 0;
  const res = await ddb.send(new UpdateItemCommand({
    TableName: TABLE, Key: { pk: { S: `day#${day}` } },
    UpdateExpression: "ADD spent :d", ExpressionAttributeValues: { ":d": { N: (actualUsd - RESERVE).toFixed(6) } }, ReturnValues: "ALL_NEW",
  }));
  return Number(res.Attributes?.spent?.N ?? 0);
}

export async function status(): Promise<{ day: string; spent: number; cap: number; remaining: number; turns: number }> {
  const day = dayKey();
  let spent = 0, turns = 0;
  if (TABLE) {
    const res = await ddb.send(new GetItemCommand({ TableName: TABLE, Key: { pk: { S: `day#${day}` } } }));
    spent = Number(res.Item?.spent?.N ?? 0); turns = Number(res.Item?.turns?.N ?? 0);
  }
  return { day, spent: Math.max(0, spent), cap: CAP, remaining: Math.max(0, CAP - spent), turns };
}
